src/target/renesas.c

```diff
@@ -0,0 +1,938 @@
+	case PNR_SERIES_RA4E1:
+	case PNR_SERIES_RA4M2:
+	case PNR_SERIES_RA4M3:
+	case PNR_SERIES_RA6M4:
+	case PNR_SERIES_RA6M5:
+	case PNR_SERIES_RA6E1:
+	case PNR_SERIES_RA6T2:
+		has_fmeprot = true;
+	default:
+		break;
+	}
+
+	if (has_fmeprot)
+		target_mem_write16(t, RV40_FMEPROT, RV40_FMEPROT_UNLOCK);
+
+	/* Set PE/READ mode */
+	uint16_t fentryr = 0;
+	switch (pe_mode) {
+	case PE_MODE_CF:
+		fentryr |= FENTRYR_PE_CF;
+		break;
+	case PE_MODE_DF:
+		fentryr |= FENTRYR_PE_DF;
+		break;
+	default:
+		break;
+	}
+	target_mem_write16(t, RV40_FENTRYR, FENTRYR_KEY | fentryr);
+
+	platform_timeout timeout;
+	platform_timeout_set(&timeout, 10);
+
+	/* Wait for the operation to complete or timeout, Read until FENTRYR and FRDY is set */
+	while (target_mem_read16(t, RV40_FENTRYR) != fentryr || !(target_mem_read32(t, RV40_FSTATR) & RV40_FSTATR_RDY)) {
+		if (target_check_error(t) || platform_timeout_is_expired(&timeout))
+			return false;
+	}
+
+	if (has_fmeprot && pe_mode == PE_MODE_READ)
+		target_mem_write16(t, RV40_FMEPROT, RV40_FMEPROT_LOCK);
+
+	return true;
+}
+
+static bool renesas_rv40_error_check(target *t, uint32_t error_bits)
+{
+	bool error = false;
+
+	uint8_t fstatr = target_mem_read32(t, RV40_FSTATR);
+
+	/* see "Recovery from the Command-Locked State": Section 47.9.3.6 of the RA6M4 manual R01UH0890EJ0100.*/
+	if (target_mem_read8(t, RV40_FASTAT) & RV40_FASTAT_CMDLK) {
+		/* if an illegal error occurred read and clear CFAE and DFAE in FASTAT. */
+		if (fstatr & RV40_FSTATR_ILGLERR) {
+			target_mem_read8(t, RV40_FASTAT);
+			target_mem_write8(t, RV40_FASTAT, 0);
+		}
+		error = true;
+	}
+
+	/* check if status is indicating a programming error */
+	if (fstatr & error_bits)
+		error = true;
+
+	if (error) {
+		/* Stop the flash */
+		target_mem_write8(t, RV40_CMD, RV40_CMD_FORCED_STOP);
+
+		platform_timeout timeout;
+		platform_timeout_set(&timeout, 10);
+
+		/* Wait until the operation has completed or timeout */
+		/* Read FRDY bit until it has been set to 1 indicating that the current  operation is complete.*/
+		while (!(target_mem_read32(t, RV40_FSTATR) & RV40_FSTATR_RDY)) {
+			if (target_check_error(t) || platform_timeout_is_expired(&timeout))
+				return error;
+		}
+
+		if (target_mem_read8(t, RV40_FASTAT) & RV40_FASTAT_CMDLK) {
+			return error;
+		}
+	}
+
+	return error;
+}
+
+static bool renesas_rv40_prepare(target_flash_s *f)
+{
+	target *t = f->t;
+
+	if (!(target_mem_read32(t, RV40_FSTATR) & RV40_FSTATR_RDY) || target_mem_read16(t, RV40_FENTRYR) != 0) {
+		DEBUG_WARN("flash is not ready, may be hanging mid unfinished command due to something going wrong, "
+				   "please power on reset the device\n");
+
+		return false;
+	}
+
+	/* code flash or data flash operation */
+	const bool code_flash = f->start < RENESAS_CF_END;
+
+	/* Transition to PE mode */
+	const pe_mode_t pe_mode = code_flash ? PE_MODE_CF : PE_MODE_DF;
+	if (!renesas_rv40_pe_mode(t, pe_mode) || renesas_rv40_error_check(t, RV40_FSTATR_ILGLERR))
```

Comment:
This looks like it can be return elided? `return renesas_rv40_pe_mode(t, pe_mode) && !renesas_rv40_error_check(t, RV40_FSTATR_ILGLERR);`